src/libtorch.cc

```diff
@@ -512,6 +581,20 @@ ModelInstanceState::ModelInstanceState(
 
 ModelInstanceState::~ModelInstanceState()
 {
+  // If weight sharing is enabled and this is the last model instance,
+  // remove the pointer from the model map
+  if (torch_model_.use_count() == 2) {
```

Comment:
Why not just let `~ModelState()` to release all the models? The instances all share the same life cycle and when Triton unloads a model, the `TRITONBACKEND_Model` will be destroyed after all `TRITONBACKEND_ModelInstance` destruction.

I don't think you need to do anything explicitly as shared_ptr will handle that for you as long as the State objects are released.

Reply:
I'm not fully following. We have two options:
- Weights are not being shared: each model is pointed to by one instance
- Weights are being shared: each model is pointed to by each instance and is in the model mapping

It seems like the ModelState is shared across devices. We cannot store the model pointers in the ModelState, since there can be multiple devices (each with their own shared pointers). And weight sharing may be disabled, in which case each  model has its own pointer.

To delete the model, we need to release all copies of the shared pointer (in this case, one in each model and one in the model mapping; that changes if we move the model pointer to ModelState).

Reply:
I understand that there will be the mapping from device to torch model, what I am saying is that you don't need to explicitly perform `map.erase()` in `ModelState` and `shared_ptr.reset()` in `ModelInstanceState`. You need to delete the `torch_model` (managed in shared_ptr) only when Triton is unloading the model, so you can expect the `ModelInstanceState` and `ModelState` objects will be destroyed.

Consider the cases of sharing / not sharing with no explicit handling:

- Sharing: at least one `ModelInstanceState` holds shared pointer of the torch model; `ModelState` has another shared pointer in the map. On unloading model, `ModelInstanceState` destruction will destroy its copy of shared pointer, and then `ModelState` destruction will destroy the map which results in destroy the shared pointer in the map. Then all shared pointers of the torch model is destroyed and should automatically result in destroying the torch model.
- Not sharing: same as current behavior.

Reply:
What is specific versions/instances of the model are being unloaded without the entire model being unloaded? For example if the user changes the instance_group field in the model config.pbtxt (with polling enabled)?

Reply:
I see, thank you for clarifying @GuanLuo.

For the edge cases @CoderHam pointed out:
- The instance_group case should work fine since the model does not change (each instance would have a pointer, the ModelState would have the same pointer).
- For versioning, each ModelState holds a different version, right? So if the model version is unloaded, the ModelState would be destroyed. If that's not the case, this PR would need to be updated to track version in the map (so model versions don't get shared).

I appreciate you thinking through the edge cases. If there's a scenario where the ModelState holds multiple models (e.g. versions) at the same time or the model gets updated without all ModelInstanceStates being destructed first, the code would need to be updated.

Reply:
Need to clarify that when it gets to the backend, "version" is no longer relevant. `TRITONBACKEND_Model` and the `TRITONBACKEND_ModelInstance` associated to it are just pointed to some files to load the model from, thus all instance will be "identical" since they are reading the model from the same set of files.

In other words, it is not the right scope to talk about "version" here, and all instances associated with a `TRITONBACKEND_Model` are simply be the copies of the same model. When instance group setting is updated (same for other file changes), think of it as creating new `TRITONBACKEND_Model` (and instances) and swap out the previous one.

Reply:
Done!